content/posts/2021-05-airdrop-anywhere-part-1.md

```diff
@@ -0,0 +1,65 @@
+
+### A first stab...
+
+AirDrop uses a protocol called Apple Wireless Direct Link (AWDL) which is used to establish a peer-to-peer wireless connection between the sender and receiver _without them being connected to the same access point_. A whole lot of people at the [Secure Mobile Networking Lab](https://github.com/seemoo-lab) have been working to reverse engineer it and that work has resulted in the [OWL project](https://owlink.org) which has a bunch of publications breaking down the protocol and its security implications as well as open source code in C, Python and Swift that provide a reference implementation of that protocol (called OWL - Open Wireless Link) and AirDrop itself.
+
+However OWL is written to work on MacOS and Linux - it makes use of `libpcap` and various libraries that only work on those platforms.
+
+Under the hood OWL uses `libpcap` to create a virtual network adapter and uses raw frame manipulation from userspace code to talk to the underlying network stack. The project has a [great diagram](https://github.com/seemoo-lab/owl#architecture) detailing their interactions with the different parts of the system:
+
+<img src="https://raw.githubusercontent.com/seemoo-lab/owl/master/resources/overview.png" width=480 alt="OWL Architecture"><br/>
+<sub style="color:lightgray">OWL Architecture</sub>
+
+To get this whole shebang working under Windows we need a similar approach of capturing and manipulating raw network packets. I recalled that [Wireshark](https://www.wireshark.org/) used to install a packet capture driver that implemented the NDIS interfaces provided by Windows. Turns out the original way it did this was with [Winpcap](https://www.winpcap.org/) but there's a note saying that it has been deprecated in favour of [npcap](https://nmap.org/npcap/) which is part of the nmap project. Sadly the licensing for npcap is just a tad expensive for any kind of distribution - like an app that could live in the Windows Store - if this project ever gets to that point then I'd probably consider writing a TAP driver much like what OpenVPN uses (see their [TAP driver repo](https://github.com/OpenVPN/tap-windows6)).
+
+I should note that, prior to reading the OWL publications, I was under the impression that this might be implemented using something like [Wintun](https://www.wintun.net). This is similar in many ways to npcap but it works uses a TUN driver rather than a TAP one. TUN drivers operate at layer 3 of the OSI stack whereas TAP drivers operate at layer 2 - [wikipedia](https://en.wikipedia.org/wiki/TUN/TAP) has a nice diagram illustrating the differences. To implement AWDL we need to be able to manipulate radiotap and 802.11 on the WLAN interface that we are connected to - these are transport layer (i.e. L2) concerns so we need to use something that operates at that layer.
```

Comment:
```suggestion
I should note that, prior to reading the OWL publications, I was under the impression that this might be implemented using something like [Wintun](https://www.wintun.net). This is similar in many ways to npcap but it works uses a TUN driver rather than a TAP one. TUN drivers operate at layer 3 of the OSI stack whereas TAP drivers operate at layer 2 - Wikipedia [has a nice diagram](https://en.wikipedia.org/wiki/TUN/TAP) illustrating the differences. To implement AWDL we need to be able to manipulate radiotap and 802.11 on the WLAN interface that we are connected to - these are transport layer (i.e. L2) concerns so we need to use something that operates at that layer.
```